### PR TITLE
Aws pipeline docker

### PR DESCRIPTION
**Pull request by pgrandinetti**

### CODE REVIEW

**Comment by pgrandinetti** [comment]
still work in progress, but can already be run and brainstormed. @RuggeroFabbiano

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+RUN apt install postgresql-server-dev-12 -y
+
+RUN apt install -y python3-venv
+RUN python3.8 -m venv /venv
+ENV PATH=/venv/bin:$PATH
+
+RUN git clone https://github.com/pgrandinetti/standard-traffic-data stdata
+RUN cd stdata \
+    && git checkout aws_pipeline \
```
Comment (pgrandinetti) [inline Dockerfile]: to remove when that branch is merged on main

**Review by RuggeroFabbiano — APPROVED** [review]

**Comment by pgrandinetti** [comment]
Just used it to generate `most_0400_1300` (7.0 Gb and 70981786 records). Worked in 2-clicks!
```
docker build -t sumotest .
docker run --env-file docker.env sumotest
```

with the `docker.env` file content as it follows:

```
# Boto secrets
AWS_KEY_ID=*****
AWS_SECRET_KEY=*****
AWS_REGION=us-east-1

# RDS
DB_USER=****
DB_PASSWORD=******
DB_PORT=5432
DB_HOST=*****
DB_NAME=*****

# Docker run options
SUMO_COMMAND=most.sumocfg -e 46800 --step-length 1 --device.fcd.period 5 --fcd-output
SUMO_MODEL_PREFIX=most
SUMO_OUTPUT_FILE=most_0400_1300_1_5
S3=1
S3_BUCKET=*****
DB=1
TABLE=most_0400_1300_1_5
DATABASE=sumodb
CLEANUP=1
```

**Comment by RuggeroFabbiano** [comment]
You are

> Just used it to generate `most_0400_1300` (7.0 Gb and 70981786 records). Worked in 2-clicks!
> 
> ```
> docker build -t sumotest .
> docker run --env-file docker.env sumotest
> ```
> 
> with the `docker.env` file content as it follows:
> 
> ```
> # Boto secrets
> AWS_KEY_ID=*****
> AWS_SECRET_KEY=*****
> AWS_REGION=us-east-1
> 
> # RDS
> DB_USER=****
> DB_PASSWORD=******
> DB_PORT=5432
> DB_HOST=*****
> DB_NAME=*****
> 
> # Docker run options
> SUMO_COMMAND=most.sumocfg -e 46800 --step-length 1 --device.fcd.period 5 --fcd-output
> SUMO_MODEL_PREFIX=most
> SUMO_OUTPUT_FILE=most_0400_1300_1_5
> S3=1
> S3_BUCKET=*****
> DB=1
> TABLE=most_0400_1300_1_5
> DATABASE=sumodb
> CLEANUP=1
> ```

What can I say, you're a rocket man!

**Comment by pgrandinetti** [comment]
Added description in README. @RuggeroFabbiano Do you want to try to reproduce a short run and see the README is clear enough? I'd like to merge this early next week.

**Review by RuggeroFabbiano — APPROVED** [review]

**Comment by RuggeroFabbiano** [comment]
> Added description in README. @RuggeroFabbiano Do you want to try to reproduce a short run and see the README is clear enough? I'd like to merge this early next week.

Sure I'll do it soon, in any case I was just about to comment how clear the read-me already seems to me!

**Comment by RuggeroFabbiano** [comment]
> Added description in README. @RuggeroFabbiano Do you want to try to reproduce a short run and see the README is clear enough? I'd like to merge this early next week.

I had to make a modification in the Dockerfile to have it working locally (see comment on file), but your instructions were crystal clear. The database upload instead rejects doesn't accept the password for user sumoreader (the one in the notebook...).

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+RUN yes Y | apt install software-properties-common
+
+RUN apt install -y git \
+    && apt install -y wget \
+    && apt install -y curl
+
+RUN add-apt-repository ppa:sumo/stable -y \
+    && apt update \
+    && apt install -y sumo sumo-tools
```
Comment (RuggeroFabbiano) [inline Dockerfile]: I had to add `--fix-missing` here to avoid a _failed to fetch_ error

**Review by RuggeroFabbiano — CHANGES_REQUESTED** [review]

**Comment by pgrandinetti** [comment]
> I had to make a modification in the Dockerfile to have it working locally (see comment on file), but your instructions were crystal clear. The database upload instead rejects doesn't accept the password for user sumoreader (the one in the notebook...).

Yes, like the name says "sumoreader" is a read-only user for the database in AWS. The pipeline writes data to the DB, hence you cannot use that user credentials. I didn't give you yet the writer user credentials.

If you set `DB=0` then it becomes irrelevant.

If you want to test the part of writing to DB, then you should use a DB in your local system (postgresql needed), and add the credentials for that local DB in the `docker.env`. The way to connect the local DB (in the host) with the docker container (the "guest") is with the option `--network host` in `docker run...`; documentation for Windows [here](https://docs.docker.com/docker-for-windows/networking/#use-cases-and-workarounds)

File: Dockerfile (hunk 2 above)
Comment (pgrandinetti) [reply Dockerfile]: This sounds like problems,I didn't get the error (in face of reproducibility).
I will look into it, in the meanwhile can you please send a PR from your local branch with the fix (any branch name is OK) to _this_ branch (`aws_pipeline_docker`)

**Comment by RuggeroFabbiano** [comment]
> > I had to make a modification in the Dockerfile to have it working locally (see comment on file), but your instructions were crystal clear. The database upload instead rejects doesn't accept the password for user sumoreader (the one in the notebook...).
> 
> Yes, like the name says "sumoreader" is a read-only user for the database in AWS. The pipeline writes data to the DB, hence you cannot use that user credentials. I didn't give you yet the writer user credentials.
> 
> If you set `DB=0` then it becomes irrelevant.
> 
> If you want to test the part of writing to DB, then you should use a DB in your local system (postgresql needed), and add the credentials for that local DB in the `docker.env`. The way to connect the local DB (in the host) with the docker container (the "guest") is with the option `--network host` in `docker run...`; documentation for Windows [here](https://docs.docker.com/docker-for-windows/networking/#use-cases-and-workarounds)

Nevermind, I did know that it was a reader only, my oversight!

**Comment by pgrandinetti** [comment]
Made some tweaks to follow the recommended best practices for apt-get. Works in my local.

@RuggeroFabbiano Can you please do again a quick test?
I would start with `docker system prune`, just to be safe.
Then `docker build` and if that doesn't work, first try with `--no-cache` in case some Windows caching is causing the problem. Docs [here](https://docs.docker.com/engine/reference/commandline/build/#options).
If that still doesn't work, then try with `--fix-missing` as you already did.

Thanks!!

**Comment by RuggeroFabbiano** [comment]
> Made some tweaks to follow the recommended best practices for apt-get. Works in my local.
> 
> @RuggeroFabbiano Can you please do again a quick test?
> I would start with `docker system prune`, just to be safe.
> Then `docker build` and if that doesn't work, first try with `--no-cache` in case some Windows caching is causing the problem. Docs [here](https://docs.docker.com/engine/reference/commandline/build/#options).
> If that still doesn't work, then try with `--fix-missing` as you already did.
> 
> Thanks!!

Test done, works like a charm. Didn't even need `--no-cache` (which I tried before `--fix-missing ` but didn't solve the issue last time).

**Review by RuggeroFabbiano — APPROVED** [review]

**Comment by pgrandinetti** [comment]
Thanks @RuggeroFabbiano  Go figure why... :)
Merging and closing

**Pull request merged**